File: backend/src/models/Payment.js
import mongoose from 'mongoose';

const paymentSchema = new mongoose.Schema(
  {
    booking_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Booking',
      required: true
    },
    amount: {
      type: Number,
      required: true,
      min: 0
    },
    currency: {
      type: String,
      default: 'ETB'
    },
    status: {
      type: String,
      enum: ['pending', 'success', 'failed'],
      default: 'pending'
    },
    chapa_tx_ref: {
      type: String,
      required: true,
      unique: true
    },
    chapa_checkout_url: {
      type: String,
      default: null
    }
  },
  {
    timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' }
  }
);

paymentSchema.set('toJSON', {
  transform: (doc, ret) => {
    ret.id = ret._id;
    delete ret._id;
    delete ret.__v;
    return ret;
  }
});

const Payment = mongoose.model('Payment', paymentSchema);
export default Payment;

File: passenger_app/lib/models/payment.dart
/// Chapa payment record — fields from OpenAPI `#/components/schemas/Payment`.
class Payment {
  final String id;
  final String bookingId;
  final double amount;
  final String currency;
  final String status;
  final String chapaTxRef;
  final String? chapaCheckoutUrl;

  const Payment({
    required this.id,
    required this.bookingId,
    required this.amount,
    required this.currency,
    required this.status,
    required this.chapaTxRef,
    this.chapaCheckoutUrl,
  });

  factory Payment.fromJson(Map<String, dynamic> json) {
    final bookingField = json['bookingId'] ?? json['booking_id'];
    final bookingId = bookingField is Map
        ? bookingField['id']?.toString() ?? ''
        : bookingField?.toString() ?? '';

    return Payment(
      id: json['id']?.toString() ?? json['paymentId']?.toString() ?? '',
      bookingId: bookingId,
      amount: (json['amount'] as num?)?.toDouble() ?? 0.0,
      currency: json['currency']?.toString() ?? 'ETB',
      status: json['status']?.toString() ?? 'pending',
      chapaTxRef:
          json['chapaTxRef']?.toString() ??
          json['chapa_tx_ref']?.toString() ??
          '',
      chapaCheckoutUrl:
          json['chapaCheckoutUrl']?.toString() ??
          json['chapa_checkout_url']?.toString(),
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'booking_id': bookingId,
    'amount': amount,
    'currency': currency,
    'status': status,
    'chapa_tx_ref': chapaTxRef,
    'chapa_checkout_url': chapaCheckoutUrl,
  };
}
